### ISSUE TITLE
Large dataset causes error in hdf5 generation during training

### ISSUE THREAD

**Issue by carlogrisetti** [uber/ludwig#992]
While training a big dataset (25GB unprocessed, I suppose around 30GB in hdf5 format), during the preprocessing phase when hdf5 generation should occur, Ludwig allocates some memory, peaks (well below any resource limit) and then spits out this message:
[CLI-output-1.txt](https://github.com/uber/ludwig/files/5484242/CLI-output-1.txt)
One CTRL+C after, this gets out
[CLI-output-2.txt](https://github.com/uber/ludwig/files/5484268/CLI-output-2.txt)
Memory goes down, then ramps up with the same pattern as before, like it tries to re-do the same data load.
One CTRL+C after, this gets out
[CLI-output-3.txt](https://github.com/uber/ludwig/files/5484270/CLI-output-3.txt)
Memory slowly ramps up again in the same fashion, reaches a peak, and the same happens again
CTRL+C and...
[CLI-output-4.txt](https://github.com/uber/ludwig/files/5484293/CLI-output-4.txt)
CTRL+C and...
[CLI-output-5.txt](https://github.com/uber/ludwig/files/5484304/CLI-output-5.txt)

After this last one, each CTRL+C returns the same exact trace as the last one. The only way to get out of this is to terminate the shell, with an increasing SpawnPoolWorker-NN number

This is the model I'm using: [model_definition_v1.yaml.txt](https://github.com/uber/ludwig/files/5484203/model_definition_v1.yaml.txt)

I also tried downgrading to tensorflow 2.2.0 and upgrading to python 3.8.6. All the traces up here have been done with tensorflow 2.3.1, python 3.8.6 and the latest ludwig from master

**Comment by tgaddair** [comment]
Looks like this issue may be related to sending large payloads back to the main thread:

https://stackoverflow.com/questions/47692566/python-multiprocessing-apply-async-assert-left-0-assertionerror

However, sounds like that specific issue should have been resolved in Python 3.8.

How large is each image in this dataset?  If you limit the number of examples to say 10%, does the issue persist?

**Comment by carlogrisetti** [comment]
A smaller dataset works, I've been training on this since last month, and I just tried expanding the dataset to around 200k samples.
Each image size varies a lot, since it is a mix of .bmp and .png, 224x224 rgb images.
Will get some statistics for you, and double-check the dataset (in the past a much bigger image slipped in the dataset, but the message was different)

**Comment by carlogrisetti** [comment]
Just retried with an older dataset of 100k samples and works like a charm.

100k samples dataset: average image size is 120kb, for 12GB total size, and an HDF5 file generated of around 14.5GB
200k samples dataset: average image size is 116kb, for 23GB total size, and an HDF5 file never gets generated, but I expect something around 25GB

Do you have any other suggestion to check?

**Comment by tgaddair** [comment]
@carlogrisetti, have you tried setting the preprocessing parameter `'in_memory': False` for the image feature?  It may be that the system just doesn't have enough RAM to read the dataset into memory.  How much RAM does this machine have?

**Comment by carlogrisetti** [comment]
56GB RAM, 100GB swapfile

I'll try anyway :)

**Comment by carlogrisetti** [comment]
100k samples work fine with or without in_memory: False
200k samples does not work with or without in_memory: False

It seems it has the same exact behavior (also because the hdf5 build seems to happen in memory no matter what)

**Comment by tgaddair** [comment]
@carlogrisetti do you know at what line in Ludwig the error is coming from?  My original suspicion was that it was coming from [here](https://github.com/uber/ludwig/blob/master/ludwig/features/image_feature.py#L298), but that cannot be the case when `in_memory: False`.  Unfortunately, the provided stack traces are outside the main thread, so it's unclear what part triggered this multiprocessing pool.

**Comment by carlogrisetti** [comment]
I tried running ludwig train with debug set, and loglevel to debug, but it does not spit out anything more. Am I doing something wrong? Can you confirm me the correct syntax?

I will try to see if I can run a debugger in there in the meantime

**Comment by tgaddair** [comment]
@carlogrisetti, can you share the config file you used for `in_memory: False`?

**Comment by carlogrisetti** [comment]
sure
[model_definition.yaml.txt](https://github.com/uber/ludwig/files/5490719/model_definition.yaml.txt)
it's basically the same, with the in_memory: False in the preprocessing section

**Comment by carlogrisetti** [comment]
Oh by the way, it gets stuck exactly [there](https://github.com/uber/ludwig/blob/master/ludwig/features/image_feature.py#L304) even if I specified in_memory False, and I checked the value for in_memory and it is obviously True like it's not being correctly loaded...
Only if I specify in_memory: false under an image: header it gets applied. (is this on purpose? I would suppose that not specifying it for a specific feature type but as a "generic" parameter it would apply to all the feature types that have that parameter available)

BTW, using an in_memory: False in the right place worked like a charm and generated the huge (29GB) hdf5 file.

**Comment by tgaddair** [comment]
Hey @carlogrisetti, looks like the config for `in_memory=False` is not specified correctly.  Can you put the preprocessing into the image feature, like this:

```
input_features:
    -
        name: ImageFile
        type: image
        encoder: resnet
        resnet_size: 8
        dropout: 0.05
        preprocessing:
            force_split: false
            in_memory: false
```

**Comment by carlogrisetti** [comment]
@tgaddair See the previous comment... if specified under an image tag (even within the preprocessing) it works as a workaround. I was indeed specifying it as a "global" preprocessing parameter, which I tought Ludwig was designed to manage, but it was not :)

**Comment by tgaddair** [comment]
Hey @carlogrisetti, sorry missed that part of the comment.  Glad it worked.  Global preprocessing parameter idea is interesting.  At the least, we want to be able to parse the config file so these sorts of errors are surface up front.

**Comment by w4nderlust** [comment]
@carlogrisetti and @tgaddair you can actuall yspecify it at a global level, but still on a per feature type basis:
```
preprocessing:
  force_split: false
  image:
    in_memory: false
```

The reason is that `force_split` is a global preprocessing parameter while `in_memory` is image specific. (so Travis' suggestions of putting `in_memory` in the feature preprocessing is correct, but `force_split` should not be placed there, it will have no effect)

The logic is: there are default preprocessing parameters, both global (force_split, split_probablities, etc.) and type specific (image.in_memory, sequence._max_sequence_length etc.). If you specify your custom ones in the preprocessing section, they override them (both global and per type), meaning that if you replace image.in_memory to be False, all image features will behave that way. The local preprocessing you specify for the specific feature will affect only that feature and not the others of the same type. So if you have 3 image inputs for instance, you can set preprocessing.image.in_memory: False to make all 3 read from disk, and if you change the preprocessing of one of them to in_memory True, only that specific one would be loaded in memory.
Hope this clarifies, probably should make it more clear in the docs too.

In conclusion: @carlogrisetti  can you try with the specified global/type preprocessing I pasted and confirm everything works fine and that we can close?

**Comment by carlogrisetti** [comment]
I already tested that and it works. The thing is that it's not a resources issue, (which on another machine got me to use in_memory from the start and works fine).

It's perfectly fine if you want to consider this a wontfix (also because it's most likely a python issue, and not a ludwig issue per se)

**Comment by w4nderlust** [comment]
I guess the issue then would be to try to figure out if there's a way to make a machine with 50gb of ram load a 30gb dataset, right?

I would say, maybe if you could figure out a synthetic dataset with the same properties as yours  by using the synthesizer (same size of images, channels and number of images), we can create the same using the same command one and try. Although right now I only have a machine with 16gb I can try to find some other way.
There could be something in the pipeline that leads to some unwanted duplication or memory waste of some kind, we would need to profile and see where it leads.

That said, I would consider this an improvement rather than an urgent bug to fix if you don't mind.

**Comment by carlogrisetti** [comment]
It is for sure not urgent, since we even have a workaround (generating the hdf5 file with the in_memory: false and then running the training in-memory).
I even thought it was not "interesting" enough for you to spend time on, that's why I was suggesting a wontfix.

An "automatic workaround" would also be to force non in-memory hdf5 generation for large datasets (maybe comparing the dataset size to the available system memory), and then resume with normal operations.

I'll see if I can manage to get a 56GB VM available for this, but it would be a Windows machine... Would that be useful?
(I would of course need to synthesize the dataset prior to that, but it should be quite easy to do, since I could then "deliver" the machine with the dataset already in place)

Let me know if this would be a viable thing for you to do. If not, as I have written above, there's an easy workaround, that could be also improved to automatically apply.

**Comment by tgaddair** [comment]
@carlogrisetti one thing worth calling out here, is that we're in the process of adding support for distributed data processing using Dask / Ray.  Once that's in place, a lot of these issues should be resolved for working with large datasets.

I think for now we can leave this as a wontfix, as you suggest.  But happy to revisit this in the future if it turns out the Dask / Ray stuff doesn't fully address these issues.

**Comment by carlogrisetti** [comment]
Dumb me, I stumbled upon the same exact problem and didn't remember this solution until Google found it :)
Search keyword was: `unable to allocate for an array with shape`

**Comment by w4nderlust** [comment]
The Dask support is there now, we have trained with datasets of billions of rows with it, check it out ;)
